File: app/src/main/java/com/aol/philipphofer/gui/sudoku/SudokuField.java
package com.aol.philipphofer.gui.sudoku;

import android.content.Context;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewStub;
import android.widget.GridLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

import com.aol.philipphofer.R;
import com.aol.philipphofer.logic.MainActivity;
import com.aol.philipphofer.logic.Position;
import com.aol.philipphofer.persistence.Data;
import com.aol.philipphofer.logic.sudoku.Number;

import java.util.Observable;
import java.util.Observer;

public class SudokuField extends GridLayout implements View.OnClickListener, Observer {

    public Position position;
    public Number number;

    private final TextView numberView;
    private GridLayout notesLayout = null;
    private TextView[] notes;

    private final MainActivity mainActivity;

    public SudokuField(Context context, AttributeSet attrs) {
        super(context, attrs);

        this.mainActivity = (MainActivity) context;

        LayoutInflater.from(context).inflate(R.layout.sudoku_field, this);

        numberView = findViewById(R.id.numberView);
        numberView.setOnClickListener(this);
    }

    /* public void init(Number number, Position position) {
        this.position = position;

        isNotes = false;
        // TODO: kann die nächste Zeile weg?
        setNumber(0);
        switchLayout(isNotes);
        error = false;

        setNumber(number);
        setBackgroundColor(ContextCompat.getColor(getContext(), R.color.unselected));
        if (number > 0) {
            setChangeable(false);
            setNumberViewText("" + number);
        } else {
            setNumberViewText("");
            setChangeable(true);
        }
    } */

    public void init(Number number, Position position) {
        this.position = position;
        this.number = number;
        this.number.addObserver(this);

        if (!number.isChangeable())
            numberView.setTypeface(numberView.getTypeface(), Typeface.BOLD);
        else
            numberView.setTypeface(numberView.getTypeface(), Typeface.ITALIC);

        this.update(this.number, null);
    }

    /* public void load(Position position) {
        this.position = position;
        setBackgroundColor(ContextCompat.getColor(getContext(), R.color.unselected));
        Data data = Data.instance(mainActivity);
        data.loadField(this);
        if (getError())
            if (Data.instance(mainActivity).loadBoolean(Data.GAME_SHOW_ERRORS))
                this.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.error));
    } */

    public void switchLayout(boolean isNotes) {
        if (isNotes) {
            this.numberView.setVisibility(INVISIBLE);
            getNotesLayout().setVisibility(VISIBLE);
        } else {
            this.numberView.setVisibility(VISIBLE);
            if (notesLayout != null)
                getNotesLayout().setVisibility(INVISIBLE);
        }
    }

    public void error(boolean error) {
        // TODO mainActivity.addError();
        if (error && Data.instance(mainActivity).loadBoolean(Data.GAME_SHOW_ERRORS))
            setBackgroundColor(ContextCompat.getColor(getContext(), R.color.error));
        else
            this.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.unselected));
    }

    public void select(boolean select) {
        if (!this.number.isError() || !Data.instance(mainActivity).loadBoolean(Data.GAME_SHOW_ERRORS)) {
            if (select)
                this.setBackgroundColor(MainActivity.getPrimaryColor(getContext()));
            else
                this.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.unselected));
        }
    }

    public void lightSelect(boolean select) {
        if (!this.number.isError() || !Data.instance(mainActivity).loadBoolean(Data.GAME_SHOW_ERRORS)) {
            if (select)
                setBackgroundColor(ContextCompat.getColor(getContext(), R.color.lightSelected));
            else
                setBackgroundColor(ContextCompat.getColor(getContext(), R.color.unselected));
        }
    }

    public void setNumberViewText(int number) {
        if (number != 0)
            this.numberView.setText(String.valueOf(number));
        else
            this.numberView.setText("");
    }

    public void setNotes(boolean[] bool) {
        for (int i = 0; i < 9; i++)
            if (bool[i])
                notes[i].setVisibility(VISIBLE);
            else
                notes[i].setVisibility(INVISIBLE);
    }

    @Override
    public void update(Observable observable, Object o) {
        Number n = (Number) observable;
        this.error(n.isError());

        // handle number
        this.setNumberViewText(n.getNumber());

        // handle notes
        this.switchLayout(n.isNotes());
        if (this.notesLayout != null)
            this.setNotes(n.getNotes());
    }

    @Override
    public void onClick(View view) {
        mainActivity.select(this.position);
    }

    private GridLayout getNotesLayout() {
        if (notesLayout == null) {
            notesLayout = (GridLayout) ((ViewStub) findViewById(R.id.notesGridStub)).inflate();
            notesLayout.setOnClickListener(this);
            notes = new TextView[9];
            for (int i = 0; i < 9; i++)
                notes[i] = findViewById(getContext().getResources().getIdentifier("tv" + i, "id", getContext().getPackageName()));
        }
        return notesLayout;
    }
}

File: app/src/main/java/com/aol/philipphofer/gui/sudoku/SudokuField.kt
package com.aol.philipphofer.gui.sudoku

import android.content.Context
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewStub
import android.widget.GridLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.aol.philipphofer.R
import com.aol.philipphofer.logic.MainActivity
import com.aol.philipphofer.logic.Position
import com.aol.philipphofer.logic.observer.Listener
import com.aol.philipphofer.logic.observer.Observable
import com.aol.philipphofer.logic.sudoku.Number
import com.aol.philipphofer.persistence.Data
import com.aol.philipphofer.persistence.Data.Constants.GAME_SHOW_ERRORS

class SudokuField(context: Context, attributeSet: AttributeSet) :
    GridLayout(context, attributeSet), OnClickListener, Listener<Number> {

    lateinit var position: Position
    lateinit var number: Number

    private val numberView: TextView
    private var notesLayout: GridLayout? = null
    private var notes: Array<TextView> = emptyArray()

    private val mainActivity: MainActivity = context as MainActivity

    init {
        LayoutInflater.from(context).inflate(R.layout.sudoku_field, this)

        numberView = findViewById(R.id.numberView)
        numberView.setOnClickListener(this)
    }

    fun init(number: Number, position: Position) {
        this.position = position

        this.number = number
        this.number.addListener(this)
        observableChanged(this.number)

        val typeFace = if (number.isChangeable) Typeface.ITALIC else Typeface.BOLD
        numberView.setTypeface(numberView.typeface, typeFace)
    }

    private fun switchLayout(isNotes: Boolean) {
        numberView.visibility = if (isNotes) INVISIBLE else VISIBLE

        if (isNotes) {
            getNotesLayout().visibility = VISIBLE
        } else {
            if (notesLayout != null)
                getNotesLayout().visibility = INVISIBLE
        }

    }

    private fun error(error: Boolean) {
        val showErrors = Data.instance(mainActivity).loadBoolean(GAME_SHOW_ERRORS)
        val color = if (error && showErrors) R.color.error else R.color.unselected

        setBackgroundColor(ContextCompat.getColor(context, color))
    }

    fun select(select: Boolean) {
        val showErrors = Data.instance(mainActivity).loadBoolean(GAME_SHOW_ERRORS)
        if (showErrors && number.isError) return

        if (select)
            this.setBackgroundColor(MainActivity.getPrimaryColor(context));
        else
            this.setBackgroundColor(ContextCompat.getColor(context, R.color.unselected));
    }

    fun lightSelect(select: Boolean) {
        val showErrors = Data.instance(mainActivity).loadBoolean(GAME_SHOW_ERRORS)
        if (showErrors && number.isError) return

        val color = if (select) R.color.lightSelected else R.color.unselected
        setBackgroundColor(ContextCompat.getColor(context, color))
    }

    private fun setNumberViewText(number: Int) {
        val stringValue = if (number != 0) number.toString() else ""
        numberView.text = stringValue
    }

    private fun setNotes(visibleNotes: BooleanArray) {
        for (i in 0..8)
            notes[i].visibility = if (visibleNotes[i]) VISIBLE else INVISIBLE
    }

    override fun onClick(view: View) {
        mainActivity.select(position)
    }

    override fun observableChanged(observable: Observable<Number>) {
        val changedNumber = observable as Number

        error(changedNumber.isError)
        setNumberViewText(changedNumber.number)
        switchLayout(number.isNotes)
        if (null != notesLayout) setNotes(changedNumber.notes)

        Data.instance(context).saveGameNumber(changedNumber, position)
    }

    private fun getNotesLayout(): GridLayout {
        if (null == notesLayout) {
            notesLayout = (findViewById<ViewStub>(R.id.notesGridStub)).inflate() as GridLayout
            notesLayout!!.setOnClickListener(this)

            notes = Array(9) {
                val id = context.resources.getIdentifier("tv$it", "id", context.packageName)
                findViewById(id)
            }
        }
        return notesLayout!!
    }
}